### PR TITLE
Fix S6964 FP: Should not raise for reference properties in nullable context

### LINKED ISSUE #9284: Fix S6964 FP: Should not raise for reference properties in nullable context
### Description
The rule [S6964](https://sonarsource.github.io/rspec/#/rspec/S6964/csharp) should not be active for model properties in a nullable aware context (where nullable reference types are enabled). This is because of:

> By default, MVC will treat a non-nullable reference type parameters and properties as-if `[Required]` has been applied, resulting in validation errors when no value was bound.

Source: https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.mvcoptions.suppressimplicitrequiredattributefornonnullablereferencetypes

Even if a project sets this property to true, the rule should not be reported in my opinion. Because then it was an explicit decision made by the developer (e.g. because the project uses a different validation framework like FluentValidation).

### Repro steps
1. Create a new ASP.NET Core Web API project (nullable reference types are enabled by default in current .NET versions)
2. Add a model with not-nullable reference type properties (e.g. `string`) and use it in a POST controller method.
3. Call the POST method without providing the property. See that ASP.NET Core returns an validation error, even though you didn't apply the `Required` attribute manually.

### Expected behavior
Rule should not be reported.

### Actual behavior
Rule is reported.

### Known workarounds
I had to create a new quality profile in SonarCloud to disable the rule completely.

### Related information

* C#/VB.NET Plugins version: 9.25.0.90414 (used by the `SonarCloudPrepare@1` task)
* Visual Studio version: not relevant
* MSBuild / dotnet version: 8.0.300
* SonarScanner for .NET version: SonarScanner for MSBuild 5.15 (used by the `SonarCloudAnalyze@1` task)
* Operating System: `windows-latest` build agent on Azure DevOps

### PR DESCRIPTION
**Pull request by mary-georgiou-sonarsource**
Fixes #9284

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/AvoidUnderPosting.cs [hunk 1]
```diff
@@ -94,8 +93,7 @@ private static void CheckInvalidProperties(INamedTypeSymbol parameterType, Sonar
         && (type.Constructors.Any(x => x.Parameters.Length == 0)        // The type must have a parameterless constructor, unless
             || type.IsValueType                                         // - it's a value type
             || type.IsRecord()                                          // - it's a record type
-            || type.IsInterface()                                       // - it's an interface (although the type that implements will be actually used)
-            || type.Is(KnownType.System_String));                       // - it has a custom Model Binder (e.g. System.String has one)
```
Comment (zsolt-kolbay-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/AvoidUnderPosting.cs:99]: The [issue title ](https://github.com/SonarSource/sonar-dotnet/issues/9284) is somewhat misleading. It not only refers to strings, but to all non-nullable reference types.
To make this work, the following statement must be removed from [this line](https://github.com/SonarSource/sonar-dotnet/blob/3fea67efaa8e1225b0c92be9ad692d89855075a5/analyzers/src/SonarAnalyzer.CSharp/Rules/AspNet/AvoidUnderPosting.cs#L100) (that checks for non-nullable reference types):
```cs
type.NullableAnnotation() != NullableAnnotation.NotAnnotated
```

File: analyzers/tests/SonarAnalyzer.Test/TestCases/AspNet/AvoidUnderPosting.CSharp8.cs [hunk 2]
```diff
@@ -6,8 +6,10 @@ namespace NullableReferences
 {
     public class ModelUsedInController
     {
+        // if a string property is not set, and nullable context is disabled there's a validation error.
+        // Otherwise, strings behave always as nullable reference types.
 #nullable enable
-        public string NonNullableReferenceProperty { get; set; }                    // Noncompliant
```
Comment (zsolt-kolbay-sonarsource) [inline analyzers/tests/SonarAnalyzer.Test/TestCases/AspNet/AvoidUnderPosting.CSharp8.cs:10]: Add an extra test case to make sure it doesn't only work for strings:
```cs
#nullable enable
        public string NonNullableReferenceProperty { get; set; }        // Compliant - ASP.NET Core treats non-nullable reference types as if they were decorated with the [Required] attribute
        public object AnotherNonNullableReferenceProperty { get; set; }
        ...
```

File: analyzers/tests/SonarAnalyzer.Test/TestCases/AspNet/AvoidUnderPosting.CSharp8.cs [hunk 3]
```diff
@@ -6,8 +6,10 @@ namespace NullableReferences
 {
     public class ModelUsedInController
     {
+        // if a string property is not set, and nullable context is disabled there's a validation error.
```
Comment (zsolt-kolbay-sonarsource) [inline analyzers/tests/SonarAnalyzer.Test/TestCases/AspNet/AvoidUnderPosting.CSharp8.cs:9]: As mentioned in a previous comment, it's not only for strings.

**Review by zsolt-kolbay-sonarsource — CHANGES_REQUESTED** [review]
Good, but changes are needed to handle all non-nullable reference types, not just strings.

File: analyzers/tests/SonarAnalyzer.Test/TestCases/AspNet/AvoidUnderPosting.CSharp8.cs [hunk 4]
```diff
@@ -23,24 +28,3 @@ public IActionResult Create(ModelUsedInController model)
         }
     }
 }
-
-namespace CustomGenerics
```
Comment (zsolt-kolbay-sonarsource) [inline analyzers/tests/SonarAnalyzer.Test/TestCases/AspNet/AvoidUnderPosting.CSharp8.cs:27]: Why were these test cases removed?

File: analyzers/tests/SonarAnalyzer.Test/TestCases/AspNet/AvoidUnderPosting.CSharp8.cs [hunk 5]
```diff
@@ -7,11 +7,16 @@ namespace NullableReferences
     public class ModelUsedInController
     {
 #nullable enable
-        public string NonNullableReferenceProperty { get; set; }                    // Noncompliant
+        public string NonNullableReferenceProperty { get; set; }
```
Comment (zsolt-kolbay-sonarsource) [inline analyzers/tests/SonarAnalyzer.Test/TestCases/AspNet/AvoidUnderPosting.CSharp8.cs:10]:
```suggestion
        public string NonNullableReferenceProperty { get; set; }  // Compliant - ASP.NET Core treats non-nullable reference types as if they were decorated with the [Required] attribute
```

**Review by zsolt-kolbay-sonarsource — APPROVED** [review]
LGTM. Please check the last two comments I've added.

File: analyzers/tests/SonarAnalyzer.Test/TestCases/AspNet/AvoidUnderPosting.CSharp8.cs (hunk 4 above)
Comment (mary-georgiou-sonarsource) [reply analyzers/tests/SonarAnalyzer.Test/TestCases/AspNet/AvoidUnderPosting.CSharp8.cs:27]: Wrong push! I'll revert.

**Pull request merged**